---
title: "Jul29_WGBS2021_Stages"
author: "James Kho"
date: "7/29/2022"
output: html_document
---
This document will go over the basic pathway for methylation analysis going from an aligned and sorted methylation bam file to Differential methylation analysis. All the process will be through the R package "methylKit".

Link to reference vignette: https://compgenomr.github.io/book/bsseq.html

Link to WGBS workshop walkthrough: 

# 1.0 Reading methylation data
```{r}
library("methylKit")
``` 

Read methylation raw data from bismark using 'processBismarkAln' function. 
*Not in chunk since it takes time to process all of the samples in. Only need to process alignment once.

methraw1 = processBismarkAln(location = "Ch01WGBS_Sorted.bam", sample.id = "Ch1WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw2 = processBismarkAln(location = "Ch02WGBS_Sorted.bam", sample.id = "Ch2WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw3 = processBismarkAln(location = "Ch03WGBS_Sorted.bam", sample.id = "Ch3WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw10 = processBismarkAln(location = "Ch010WGBS_Sorted.bam", sample.id = "Ch10WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw11 = processBismarkAln(location = "Ch011WGBS_Sorted.bam", sample.id = "Ch11WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw12 = processBismarkAln(location = "Ch012WGBS_Sorted.bam", sample.id = "Ch12WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw28 = processBismarkAln(location = "Ch028WGBS_Sorted.bam", sample.id = "Ch28WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw29 = processBismarkAln(location = "Ch029WGBS_Sorted.bam", sample.id = "Ch29WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw30 = processBismarkAln(location = "Ch030WGBS_Sorted.bam", sample.id = "Ch30WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw42 = processBismarkAln(location = "Ch042WGBS_Sorted.bam", sample.id = "Ch42WGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")
methraw1A = processBismarkAln(location = "Ch1AWGBS_Sorted.bam", sample.id = "Ch1AWGBS", assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", read.context = "CpG", save.folder = "methylkit")


```{r}
## Create methylkit object
# Next let's make a file list showing the path to the processed files!
file.list2 = list (file.path("methylkit", "Ch13WGBS_CpG.txt"), 
                   file.path("methylkit", "Ch14WGBS_CpG.txt"), 
                   file.path("methylkit", "Ch15WGBS_CpG.txt"),
                   file.path("methylkit", "Ch43WGBS_CpG.txt"), 
                   file.path("methylkit", "Ch45WGBS_CpG.txt")) 

# Now we need to read the files using 'methRead'. Each file contains the methylation calls for a given sample
myobj3 = methRead(file.list2, sample.id = list("Ch13WGBS", "Ch14WGBS", "Ch15WGBS", "Ch43WGBS", "Ch45WGBS"), assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", treatment = c(1, 1, 1, 2, 2), context = "CpG", mincov = 10)
```

myobj4 = methRead(file.list2, sample.id = list("Ch4WGBS", "Ch5WGBS", "Ch6WGBS", "Ch22WGBS", "Ch23WGBS", "Ch24WGBS"), assembly = "GCF_900700415.1_Ch_v2.0.2_genomic", treatment = c(1, 1, 1, 2, 2, 2), context = "CpG", dbtype = "tabix", dbdir = "methylDB", mincov = 10)

# 2.0 MethylKit quality control
```{r}
# At this point we can also do quality check before proceeding with the rest of the analyses. There are two different QCs that I recommend.The first is methylation distribution.

getMethylationStats(myobj3[[2]], plot=TRUE, both.strands=FALSE) #Replace [[2]] with sample of interest!

#This looks at the distribution of methylation values. We expect a bimodal distribution where most CpGs have either high methylation or low methylation. Lack of bimodal suggests poor experimental quality including issues with bisulfite treatments. 

#Next we have CpG coverage. Bases with unusually high coverage are of concern as it can indicate PCR bias. Here we use 'getCoverageStats'.

getCoverageStats(myobj3[[2]], plot=TRUE, both.strands=FALSE) #Again, replace [[2]] with sample of interest.

#If we need to filter samples based on read coverage, then use 'filterByCoverage'. The function below discards bases that have 10x> coverage and those that have more than 99.9th percentile of coverage in each sample.
filtered.myobj=filterByCoverage(myobj3,lo.count=10,lo.perc=NULL,hi.count=NULL,hi.perc=99.9)

#The vignette also recommends normalizing coverage distributions across samples, which is always a great idea so below is the code:
normalized.myobj=normalizeCoverage(filtered.myobj)

# After looking through all the QC filters, next we want to merge the samples into one data object using the 'unite' function that creates a new object using the CpGs covered in each sample.
meth = unite(normalized.myobj, destrand = TRUE, save.db = TRUE)
```

Next we filter CpGs. The goal is to remove CpGs that are not variable before doing DM analysis as well as removing Cs that are potentially C -> T mutations (if known! I don't have that info for my samples so I will skip it). Let's do filtering based on variation first:

```{r}
pm = percMethylation(meth) # Get percent methylation matrix first.

mds = matrixStats::rowSds(pm) # Calculate standard deviation of CpGs

head(meth[mds>10,]) # Filter based on standard deviation.

hist(mds, col="cornflowerblue", xlab="Std. dev. per CpG")
```

#3.0 Clustering samples

After filtering, we can use 'clusterSamples' to group data points by their similarity. This can be skipped depending on what you need as the PCA will show a similar pattern (or should).

```{r}
# This function will cluster samples and draw a dendogram. It uses correlation distance (more info in vignette). The tree will be drawn using the "ward" method or "bottom up" approach. 

clusterSamples(meth, dist="correlation", method="ward.D2", plot=TRUE)

hc=clusterSamples(meth, dist="correlation", method="ward.D2", plot=FALSE)

```
We can also do PCA analysis on our samples. First let's plot for importance of components, then plot PC1 and PC2.

```{r}
PCASamples(meth, screeplot = TRUE)

PCASamples(meth)
```

#4.0 Differential methylation analysis

Now we will look for differential methylated CpG sites (DMCs) and regions (interesting regions or DMRs). The function used here will be 'calculateDiffMeth' which employs Fisher's exact test to compare test samples to a control or to another treatment. You can pool samples from a treatment together but it will not take into account biological variability between replicates.

After calculating the P-values for all bases using 'calculateDiffMeth', the next step is to filter to get differentially methylated CpGs. This is done via the 'getMethylDiff' function or filtering through '[]' function. Here I used 'getMethylDiff' as it was showcased in the workshop I attended. See vignette for example of '[]' usage.

```{r}
#Perform differential methylation analysis
myDiff = calculateDiffMeth(meth)

myDiff25p.hyper=getMethylDiff(myDiff,difference=25,qvalue=0.01,type="hyper")
myDiff25p.hypo=getMethylDiff(myDiff,difference=25,qvalue=0.01,type="hypo")
myDiff25p=getMethylDiff(myDiff,difference=25,qvalue=0.01)

bedgraph(myDiff25p.hyper, file.name = "Sdev23_T13_12h_CpG_hyper.bedgraph", col.name = "qvalue")
bedgraph(myDiff25p.hypo, file.name = "Sdev23_T13_12h_CpG_hypo.bedgraph", col.name = "qvalue")
bedgraph(myDiff25p, file.name = "Sdev23_T13_12h_CpG_all.bedgraph", col.name = "qvalue")

DiffMethPerChrS2 = diffMethPerChr(myDiff,plot=FALSE,qvalue.cutoff=0.01, meth.cutoff=25)

write.csv(DiffMethPerChrS2, file = "DiffMethPerChr_Sdev23_T13_12h.csv")
```


###myDiff.all = getMethylDiff(myDiff, qvalue = 0.05, difference = 10, type = "all") #You can change the qvalue and methylation difference (percentage) to fit your data.
bedgraph(myDiff.all, file.name = "Sdev23_T11_12h_all_CpG.bedgraph", col.name = "qvalue") # Generate Bedgraphs for visualization in IGV or other software!

myDiff.hyper2 = getMethylDiff(myDiff, qvalue = 0.05, difference = 10, type = "hyper") 
bedgraph(myDiff.hyper2, file.name = "Sdev23_T11_6h_hyper.CpG2.bedgraph", col.name = "qvalue")
###

myDiff.hypo2 = getMethylDiff(myDiff2, qvalue = 0.05, difference = 10, type = "hypo")
bedgraph(myDiff.hypo2, file.name = "stages2-3_hypo.CpG2.bedgraph", col.name = "qvalue")


Next we do the same thing but for DMRs. To do this we have to first bin the results using "tileMethylCounts" and specify the size of the region. It's important to look for DMRs as not all the CpGs will be covered in all samples at a base-pair level (i.e. base pairs with low coverage can skew the power of tests).

Here we tile the whole genome to tiles with predefined lengths but you can also use pre-defined regions such as promoters or CpG islands if known. This is done by adding up C and T counts from each covered cytosine and returning a total C and T count for each region (from vignette). Info on looking at particular regions is found on vignette but you use the R pacakage 'genomation' to provide the location of specific regions. 
```{r}
#Bin results to find DMRs
tiles2 = tileMethylCounts(normalized.myobj, win.size = 1000, step.size = 1000, cov.bases = 10)

meth.tiles = unite(tiles2, destrand = TRUE, save.db = TRUE)

myDiff.tiles2 = calculateDiffMeth(meth.tiles)
myDiff.tiles2.all = getMethylDiff(myDiff.tiles2, qvalue = 0.1, difference = 10, type = "all")
bedgraph(myDiff.tiles2.all, file.name = "Sdev23_T13_12h_DMR_All.bedgraph", col.name = "qvalue")

myDiff.tiles2.hyper = getMethylDiff(myDiff.tiles2, qvalue = 0.1, difference = 10, type = "hyper")
bedgraph(myDiff.tiles2.hyper, file.name = "Sdev23_T13_12h_DMR_Hyper.bedgraph", col.name = "qvalue")


myDiff.tiles2.hypo = getMethylDiff(myDiff.tiles2, qvalue = 0.1, difference = 10, type = "hypo")
bedgraph(myDiff.tiles2.hypo, file.name = "Sdev23_T13_12h_DMR_Hypo.bedgraph", col.name = "qvalue")
```


#Finally, let's generate some PCAs!
PCA for CpGs
```{r}
PCASamples(meth, screeplot = FALSE, adj.lim = c(1, 1), scale = TRUE, center = TRUE, comp = c(1,2), transpose = TRUE, sd.filter = TRUE, sd.threshold = 0.5, filterByQuantile = TRUE, obj.return = TRUE, chunk.size = 1e+06)
```
PCoA_Sdev23_T13_12h_CpG

PCA for DMRs:
```{r}
PCASamples(meth.tiles, screeplot = FALSE, adj.lim = c(1, 1), scale = TRUE, center = TRUE, comp = c(1,2), transpose = TRUE, sd.filter = TRUE, sd.threshold = 0.5, filterByQuantile = TRUE, obj.return = TRUE, chunk.size = 1e+06)
```
PCoA_Sdev23_T13_12h_DMR

To remove individual environment files use this:

rm(myobj3)
rm(myDiff25p.hyper)
rm(myDiff25p.hyporm)
rm(myDiff.tiles2)
rm(myDiff.tiles2.hyper)
rm(myDiff.tiles2.hypo)
rm(myDiff2)
rm(tiles2)
rm(meth2)
rm(meth.tiles2)
rm(file.list2)

#CpG Annotations
Figuring out if differentially methylated regions/bases are in genes/gene parts/regulatory regions. We need the gene annotation info from BED file (obtained via UCSC) and mydiff files: myDiff25p.hyper, myDiff25p.hypo, myDiff25p.This annotation will tell us what percentage of our DMR are on promoters/introns/exons/intergenic region.

```{r}
library(genomation)

#Let's load the annotation data i.e. coordinates of promoters, TSS, intron and exons
#I moved the annotation files into the extdata folder of my methylkit package folder. The remove.unusual = FALSE is used because my chromosome names are not standard/typical.
gene.obj <- readTranscriptFeatures(system.file("extdata", "Charengus_Geneannotation_BED.txt", package = "methylKit"), remove.unusual = FALSE)

#Next, I annotate the DM CpGs with the promoter/exon/intron using genomation functions (which works on GRanges objects, so we first coerce methylKit objects to GRanges).
annotateWithGeneParts(as(myDiff25p, "GRanges"), gene.obj)
```

We can read CpG island annotation and annotate our DMS/R with them as well. First let's read the shores and flanking regions and name the flanks as shores and CpG islands as CpGi.

```{r}
cpg.obj <-  readFeatureFlank(system.file("extdata", "Charengus_CpGannotation_BED.txt", package = "methylKit"), feature.flank.name = c("CpGi", "shores"), remove.unusual = FALSE)

#Next, I converted methylDiff object to GRanges and annotate.
diffCpGann = annotateWithFeatureFlank(as(myDiff25p, "GRanges"), cpg.obj$CpGi, cpg.obj$shores, feature.name = "CpGi", flank.name = "shores")
```

#Regional analysis
We can summarize methylation info over a set of defined regions (promoters/cpg islands) and then outputting methylRaw or methylRawList object depending on input. We will use the annotated output from the above genomation functions for promoters. For regional summary functions, we need to provide regions of interest as GRanges object.

```{r}
promoters = regionCounts(myobj3, gene.obj$promoters)

head(promoters[[1]])

```

#Other annotation objects uses
After getting annotation of DMR/DMS, we can get distance to TSS and nearest gene name using the below function:
```{r}
diffAnn = annotateWithGeneParts(as(myDiff25p, "GRanges"), gene.obj)

TSS = getAssociationWithTSS(diffAnn)
write.csv(TSS, "Sdev23_T13_12h_TSS.csv")
```

We can also get the percentage/number of DMR that overlap with intron/exon/promoters.

```{r}
getTargetAnnotationStats(diffAnn, percentage = TRUE, precedence = TRUE)

#Then we can plot this.
plotTargetAnnotation(diffAnn, precedence = TRUE, main = "Differential Methylation Annotation")

```

Sdev23_T13_12h_DiffMethAnnotation

We can then plot CpG island annotation in the same way. Plot below will show percent of diff methylated bases that are on CpG islands, CpG island shores, and other regions.

```{r}
plotTargetAnnotation(diffCpGann, col=c("green", "blue", "grey"), main = "Differential Methylation Annotation CpG")
```

Sdev23_T13_12h_DiffMethCpGAnnotation

Finally, it might also be useful to get percentage of intron/exon/promoters that overlap with differentially methylated bases.

```{r}
getFeatsWithTargetsStats(diffAnn, percentage = TRUE)
```

To Export mehtylkit obj (methyldiffDB for example), you can use the getData function to output the table.
```{r}
#myDiffData <- getData(myDiff)
#write.csv(myDiffData, "Sdev12_T11_12h_CpG_List.csv")
#But this returns a large table that may be redundant. Instead try grabbing the promoters for example or TSS
```

#Hyper/Hyposcripts:
Hyper first:
```{r}
diffHyperAnn = annotateWithGeneParts(as(myDiff25p.hyper, "GRanges"), gene.obj)
TSS_Hyper = getAssociationWithTSS(diffHyperAnn)
write.csv(TSS_Hyper, "Sdev23_T13_12h_HyperTSS.csv")

plotTargetAnnotation(diffHyperAnn, precedence = TRUE, main = "Differential Methylation Annotation")
```
Sdev23_T13_12h_DiffHyperAnnotation

```{r}
diffhyperCpGann = annotateWithFeatureFlank(as(myDiff25p.hyper, "GRanges"), cpg.obj$CpGi, cpg.obj$shores, feature.name = "CpGi", flank.name = "shores")

plotTargetAnnotation(diffhyperCpGann, col=c("green", "blue", "grey"), main = "Differential Methylation Annotation CpG")
```
Sdev23_T13_12h_DiffHyperCpGAnnotation

Now Hypo:
```{r}
diffHypoAnn = annotateWithGeneParts(as(myDiff25p.hypo, "GRanges"), gene.obj)
TSS_Hypo = getAssociationWithTSS(diffHypoAnn)
write.csv(TSS_Hypo, "Sdev23_T13_12h_HypoTSS.csv")

plotTargetAnnotation(diffHypoAnn, precedence = TRUE, main = "Differential Methylation Annotation")
```
Sdev23_T13_12h_DiffHypoAnnotation


```{r}
diffhypoCpGann = annotateWithFeatureFlank(as(myDiff25p.hypo, "GRanges"), cpg.obj$CpGi, cpg.obj$shores, feature.name = "CpGi", flank.name = "shores")

plotTargetAnnotation(diffhypoCpGann, col=c("green", "blue", "grey"), main = "Differential Methylation Annotation CpG")
```
Sdev23_T13_12h_DiffMethHypoAnnotation